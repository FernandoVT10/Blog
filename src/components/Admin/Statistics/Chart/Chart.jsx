import { useEffect, useRef } from "react";
import Api from "../../../../ApiController";

export default ({ type }) => {
    const chart = useRef();

    useEffect(() => {
        const ctx = document.getElementById("viewsChart").getContext("2d");

        Chart.defaults.global.elements.point.borderWidth = 8;
        Chart.defaults.global.defaultFontSize = 18;
        Chart.defaults.global.defaultFontColor = "white";
        Chart.defaults.global.legend.display = false;
        Chart.defaults.global.maintainAspectRatio = false;

        chart.current = new Chart(ctx, {
            type: "line",
            data: {
                labels: ["January", "February", "March", "April", "May"],
                datasets: [
                    { 
                        data: [0, 0, 0, 0, 0],
                        label: "Views",
                        borderColor: "#8BD1FF",
                        fill: false
                    }
                ]
            }
        });
    }, []);

    useEffect(() => {
        Api.get(`views/getViewsHistory/${type}/5`, true)
        .then(res => {
            const labels = res.map(({ name }) => name);
            const data = res.map(({ views }) => views);

            // Here we turn the arrays because the data is ordered by date
            labels.reverse();
            data.reverse();

            chart.current.data.labels = labels;
            chart.current.data.datasets[0].data = data;
            chart.current.update();
        });
    }, [type]);

    return (
        <canvas
        id="viewsChart"
        className="statistics-views__chart"
        width="2400"
        height="400"></canvas>
    );
};