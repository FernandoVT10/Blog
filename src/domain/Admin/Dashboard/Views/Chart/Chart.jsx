import ApiController from "../../../../../services/ApiController";

import { useEffect, useRef } from "react";

export default ({ type }) => {
    const chart = useRef();

    useEffect(() => {
        const ctx = document.getElementById("viewsChart").getContext("2d");

        Chart.defaults.global.elements.point.borderWidth = 8;
        Chart.defaults.global.defaultFontSize = 18;
        Chart.defaults.global.defaultFontColor = "white";
        Chart.defaults.global.legend.display = false;
        Chart.defaults.global.maintainAspectRatio = false;

        chart.current = new Chart(ctx, {
            type: "line",
            data: {
                labels: ["January", "February", "March", "April", "May"],
                datasets: [
                    { 
                        data: [0, 0, 0, 0, 0],
                        label: "Views",
                        borderColor: "#8BD1FF",
                        fill: false
                    }
                ]
            }
        });
    }, []);

    useEffect(() => {
        ApiController.get(`articles/views?type=${type}&limit=5`, true)
        .then(res => {
            if(res.data) {
                const { views } = res.data;
                const labels = [];
                const data = [];

                views.forEach(({ name, views }) => {
                    labels.push(name);
                    data.push(views);
                });

                // Here we turn the arrays because the data is ordered by date
                labels.reverse();
                data.reverse();

                chart.current.data.labels = labels;
                chart.current.data.datasets[0].data = data;
                chart.current.update();
            }
        });
    }, [type]);

    return (
        <canvas
        id="viewsChart"
        className="dashboard-views__chart"
        width="2400"
        height="400"></canvas>
    );
};