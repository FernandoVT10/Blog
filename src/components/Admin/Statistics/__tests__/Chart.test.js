import { act } from "react-dom/test-utils";
import { render } from "react-dom";
import Chart from "../Chart/";

const MOCK_DAY_VIEWS = [
    {
        _id: 1,
        title: "Day view mock #1",
        cover: "day-cover-1.jpg",
        views: 3951
    },
    {
        _id: 2,
        title: "Day view mock #2",
        cover: "day-cover-2.jpg",
        views: 1915
    },
    {
        _id: 3,
        title: "Day view mock #1",
        cover: "day-cover-1.jpg",
        views: 1422
    }
];
const MOCK_MONTH_VIEWS = [
    {
        _id: 1,
        title: "Month view mock #1",
        cover: "month-cover-1.jpg",
        views: 3951
    },
    {
        _id: 2,
        title: "Month view mock #2",
        cover: "month-cover-2.jpg",
        views: 1915
    },
    {
        _id: 3,
        title: "Month view mock #1",
        cover: "month-cover-1.jpg",
        views: 1422
    }
];

// Mock Chart js

class ChartMock {
    static defaults = {
        global: {
            elements: {
                point: {}
            },
            legend: {}
        }
    };

    constructor(_, options) {
        this.data = options.data;
    }

    update() {}
}

window.Chart = ChartMock;

let container;

describe('<Chart/> Component', () => {
    beforeEach(() => {
        fetchMock.doMock();

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it("Check if renders correctly", async () => {
        fetchMock.mockResponse(JSON.stringify(MOCK_DAY_VIEWS));

        document.getElementById = () => ({ getContext: jest.fn() });

        await act(async () => {
            render(<Chart/>, container);
        });

        expect(container).toMatchSnapshot();
    });
});