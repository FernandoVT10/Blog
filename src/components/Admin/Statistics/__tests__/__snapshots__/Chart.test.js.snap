// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart/> Component Check if renders correctly 1`] = `
<div>
  <canvas
    class="statistics-views__chart"
    height="400"
    id="viewsChart"
    width="2400"
  />
</div>
`;
